toolkit/tools/internal/pkggraph/cyclefind.go

```diff
@@ -0,0 +1,145 @@
+		make([]int64, 0),
+	}
+
+	// Create a temporary root node, by using a constant goalNodeName, it will act as a mutex against concurrent
+	// cycle searches on a given graph as this below call will fail if there is already a goal node with the same value.
+	rootNode, err := g.AddGoalNode(goalNodeName, nil, false)
+	if err != nil {
+		return
+	}
+
+	// This call will also remove all edges connected to the temporary root node.
+	defer g.RemoveNode(rootNode.ID())
+
+	// Seed the initial metadata state
+	metadata.parent[rootNode.ID()] = -1
+	metadata.state[rootNode.ID()] = unvisited
+
+	foundCycle, err := cycleDFS(g, rootNode.ID(), &metadata)
+	if err != nil {
+		return
+	}
+
+	if foundCycle {
+		// Convert the slice of node IDs to references to the actual nodes.
+		for _, id := range metadata.cycle {
+			nodes = append(nodes, g.Node(id).(*PkgNode).This)
+		}
+	}
+
+	return
+}
+
+// cycleDFS implements a custom DFS that updates metaData.cycle with the first cycle it finds in a given graph.
+func cycleDFS(g *PkgGraph, rootID int64, metaData *dfsData) (foundCycle bool, err error) {
+	// Recursing on a node that has already been visited indicates a fatal error with the search.
+	if metaData.state[rootID] != unvisited {
+		err = fmt.Errorf("node %d is in a bad state (%d)", rootID, metaData.state[rootID])
+		return
+	}
+
+	// Mark that rootID is actively being searched ("inProgress").
+	//
+	// If all of its neighbors have been recursively exhausted then rootID will be marked as "done"
+	// to indicate that no cycle was found in all walks containing rootID.
+	//
+	// However, if any neighbors of rootID are currently marked as "inProgress" this indicates that
+	// the current walk has encountered a cycle.
+	//
+	// Example:
+	//    a → c → a
+	//    ↓   ↓
+	//    b   b
+	//
+	// 1) visit(a)
+	//   - mark "a" as "inProgress"
+	//   - neighbor "b" is "unvisited", visit.
+	// 2) visit(b)
+	//   - mark "b" as "inProgress"
+	//   - no neighbors
+	//   - mark "b" as "done"
+	// 3) Resume visit(a)
+	//   - neighbor "c" is "unvisited", visit.
+	// 4) visit(c)
+	//   - mark "c" as "inProgress"
+	//   - neighbor "b" is "done", skip.
+	//   - neighbor "a" is "inProgress", thus a cycle between "c" and "a" is found.
+
+	metaData.state[rootID] = inProgress
+
+	for _, neighbor := range graph.NodesOf(g.From(rootID)) {
+		v := neighbor.ID()
+		if _, exists := metaData.state[v]; !exists {
+			metaData.state[v] = unvisited
+		}
+
+		switch metaData.state[v] {
+		case done:
+			continue
+		case unvisited:
+			metaData.parent[v] = rootID
+			foundCycle, err = cycleDFS(g, v, metaData)
+			if err != nil || foundCycle {
+				return
+			}
+		case inProgress:
+			createCycle(g, metaData, rootID, v)
+			foundCycle = true
+			return
+		default:
+			err = fmt.Errorf("node %d is in a bad state (%d)", v, metaData.state[v])
+			return
+		}
+	}
+
+	metaData.state[rootID] = done
+	return
+}
+
+// updateCycle records the cycle between startID and endID in metaData.cycle.
```

Comment:
Name miss-match between comment and function.
```suggestion
// createCycle records the cycle between startID and endID in metaData.cycle.
```